src/core/reference/include/openvino/reference/multinomial.hpp

```diff
@@ -0,0 +1,176 @@
+    }
+
+    // Create a cdf of probabilties on the last axis, per batch. Note cumsum exclusive  == false
+    std::vector<T> cdf(total_inputs_elements_count);
+    auto last_axis = probs_shape.size() - 1;
+    cumsum<T, size_t>(input_vals.data(), last_axis, cdf.data(), probs_shape, false, false);
+
+    // Obtain max value from cdf, per batch (from cumsum it is the last element)
+    std::vector<T> max_value_per_batch(total_inputs_elements_count / probs_shape[last_axis]);
+    Shape max_value_per_batch_shape(probs_shape);
+    max_value_per_batch_shape[last_axis] = 1;
+    std::vector<int64_t> start{static_cast<int64_t>(probs_shape[last_axis] - 1)};
+    std::vector<int64_t> step{1};
+    std::vector<int64_t> target_axis_vec{static_cast<int64_t>(last_axis)};
+    slice(reinterpret_cast<const char*>(cdf.data()),
+          probs_shape,  // == cdf shape
+          reinterpret_cast<char*>(max_value_per_batch.data()),
+          max_value_per_batch_shape,
+          sizeof(T),
+          start,
+          step,
+          target_axis_vec);
+
+    // Normalize the cdf by dividing all elements by the max value in each batch
+    std::vector<T> max_value_per_batch_divisor(total_inputs_elements_count);
+    ov::AxisSet target_axis_set = ov::AxisSet({last_axis});
+    broadcast(reinterpret_cast<const char*>(max_value_per_batch.data()),
+              reinterpret_cast<char*>(max_value_per_batch_divisor.data()),
+              max_value_per_batch_shape,
+              probs_shape,  // expand to original shape (expands last dim)
+              target_axis_set,
+              sizeof(T));
+    divide(reinterpret_cast<const T*>(cdf.data()),
+           reinterpret_cast<const T*>(max_value_per_batch_divisor.data()),
+           reinterpret_cast<T*>(cdf.data()),
+           total_inputs_elements_count,
+           false);
+
+    // Generate random probability samples
+    std::vector<double> uniform_samples(total_output_elements_count);
+    const double zero = 0;
+    const double one = 1;
+    ov::Shape output_shape_shape{output_shape.size()};
+    std::pair<uint64_t, uint64_t> initial_state(0, 0);
+    random_uniform(output_shape.data(),
+                   reinterpret_cast<const char*>(&zero),
+                   reinterpret_cast<const char*>(&one),
+                   reinterpret_cast<char*>(uniform_samples.data()),
+                   output_shape_shape,
+                   ov::element::f64,
+                   global_seed,
+                   op_seed,
+                   initial_state);
+
+    auto first_dim_size         = probs_shape.size() == 2 ? probs_shape[0]          : (size_t)1;
+    auto second_input_dim_size  = probs_shape.size() == 2 ? probs_shape[1]          : probs_shape[0];
+    auto second_output_dim_size = probs_shape.size() == 2 ? (size_t)num_samples[0]  : probs_shape[0];
+
+    // Iterate over each channel in uniform samples
+    test<T>(cdf.data(), probs_shape, &test_nr);
```

Comment:
```suggestion
```
Is it debug function? Should be removed?

Reply:
Removed!